text/0141-linking-traces.md

```diff
@@ -0,0 +1,367 @@
+
+```TypeScript
+function startSpan(options: StartSpanOptions);
+
+interface StartSpanOptions: {
+  //... other options (name, attributes, etc)
+  links?: Link[];
+}
+```
+
+SDKs should follow the OpenTelemetry spec for the [`Link` interface](https://github.com/open-telemetry/opentelemetry-js/blob/main/api/src/trace/link.ts) as defined by the platform. Non-Otel SDKs should orient themselves on Otel, resulting in the following interface:
+
+```typescript
+interface Link {
+  // contains the SpanContext of the span to link to
+  context: SpanContext;
+  // key-value pair with primitive values
+  attributes?: Attributes;
+}
+
+
+// see https://github.com/open-telemetry/opentelemetry-js/blob/main/api/src/trace/span_context.ts
+// or interface of respective platform
+interface SpanContext {
+  traceId: string,
+  spanId: string,
+  traceFlags: string,
+}
+
+type Attributes = Record<string, AttributeValues>
+type AttributeValues = string | boolean | number | Array<string> | Array<boolean> | Array<number>
+```
+
+Note: On some platforms, the Otel `Link` interface exposes another optional property: `droppedAttributesCount`. POtel SDKs should support passing in this property as defined by the API but can further ignore it when serializing the span link to Sentry envelopes.  In [JS for example](https://github.com/open-telemetry/opentelemetry-js/blob/main/api/src/trace/link.ts), the `droppedAttributeCount` can be passed, while [Python](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/trace/span.py#L120) does not permit it. 
+Non-Otel SDKs are free to ignore this property.
+
+Note II: POtel SDKs today already have to expose `addLink(s)` APIs, however we simply disregard the added links when serializing Otel spans to transaction event envelopes. Furthermore, for platform agnostic APIs, the non-POtel browser SDKs today also expose these methods. They simply no-op at the moment.
+
+#### Usage Example
+
+Adding span links should be possible at span start time, as well as when holding a reference to the span:
+
+```typescript
+// 1st trace starts
+const pageloadSpan = startInactiveSpan(...)
+
+// 2nd trace starts
+const navigation1Span = startInactiveSpan({
+  name: '/users', 
+  links: [{
+    context: pageloadSpan.spanContext(), 
+    attributes: {
+      'sentry.link.type': 'previous_trace'
+    }
+  }]
+});
+
+// 3rd trace starts
+const navigation2Span = startSpan({name: '/users/:id'}, (span) => {
+  span.addLink({
+    context: navigation1Span.spanContext(), 
+    attributes: {
+      'sentry.link.type': 'previous_trace'
+    }
+  })
+})
+```
+
+In this example, by adding span links, we can link from the last navigation trace all the way back to the initial pageload trace. By passing the `'sentry.link.type': 'previous_trace'` attribute, we can identify the link as a previous trace link in Sentry and display the spans accordingly. 
```

Comment:
This example talks about navigation traces, but what about other kinds of traces? Especially once traces are shortened? You can imagine having traces for interactions, web requests, etc - is the idea still to link each of these to the previous one? What about when the previous trace is still active when the new one starts? e.g.

```
[-----------trace 1, span A: interaction-----------]
               [ trace 1, span B: http request ]
   
     [trace 2, span C: http request]
     
                                                         [trace 3, span D: http request]
```
Does trace 2 link to trace 1 as the previous trace, or whatever the previous finished trace was? (Therefore forming a tree rather than a list). Does trace 3 then link to trace 2, even though trace 1 is the most recent span to finish? etc.

Reply:
If the crappy ASCII isn't clear, to rephrase in text:

1. Trace 1 starts
2. Trace 2 starts
3. Trace 2 ends
4. Trace 1 ends
5. Trace 3 starts
6. Trace 3 ends

How are these three traces joined by links?

Reply:
What about changing the relationship of `previous_trace` from 1:1 to 1:n and renaming it to `previous_traces`?

Reply:
I think whether turning linked transactions into a graph instead of a list makes sense partly depends on the product goals, since we haven't really outlined exactly what this is supposed to mean in the product. (As Lukas mentioned in another comment, that should be coming soon).

Reply:
I guess IOW, I would err towards accepting that linking back to previous traces might not be ideal for overlapping or parallel running traces. 

We should check though if this is a larger concern in mobile SDKs. @philipphofmann do you have insights on this?

Reply:
> However, while possible, we don't document overlapping traces today and we don't consider overlapping traces as "specc'd behavior". Setting this up in the browser for instance, would require significant manual intervention by users and I'm not aware of a lot of cases where this might happen.

FWIW this is defined and normal behaviour in OTel, so we will have to cross this bridge under OTLP if not sooner. (Imagine someone who instrumented their HTTP client spans in web, and nothing else: every one would start a trace, overlapping whenever the HTTP requests overlap.)
 
Given this proposal is primarily aimed at Sentry SDK users, and if our SDK won't have separate traces for things like parallel interactions or anything either, then I agree it is not currently a problem and so this alone shouldn't block anything here. 👍

Reply:
> What about changing the relationship of previous_trace from 1:1 to 1:n and renaming it to previous_traces?

sorry @philipphofmann almost missed this! My thinking is that `"previous_trace"` is fine because a span can have multiple links where each link has a `sentry.link.type: "previous_trace"` attribute. Does this work or am I missing something?

Reply:
I added a section about how SDKs should set the previous trace relation ship span link